### ISSUE TITLE
Get a hashable query expression text

### ISSUE THREAD

**Issue by SteffenMangold** [dotnet/efcore#14230]
By now it is extremely difficult to generate a hashable expression from an `IQueryable` object.

Hashable expressions are important for caching like in the popular extension [EFSesondLevelCaching](https://github.com/VahidN/EFSecondLevelCache.Core). This is one of the most popular and a recommended extensions by Microsoft.

Most implementations for caching are using a `ToSql` extension method [like this one](https://stackoverflow.com/a/51583047/1112048).

The main problem here is that this casting and reflection monster is so extreme slow, that if you use the cache on a high frequency it is really a bad CPU/performance killer.

An alternative is the using of `query.Expression.ToString()`. At first glance, this one generates a really nice readable expression text that looks perfect for caching.
But the problem of closure references hits you hard.
The problem is [descript here in perfect details](https://petemontgomery.wordpress.com/2008/08/07/caching-the-results-of-linq-queries/).
Downside here is that its also a heavy implementation and has some issues on special cases.

So my question(s) is(are), with you way more deeply knowledge of the EF core Source.

Is there a fast and natural way to get an expression text including resolved referencial parameters?
or
If not is that something you could consider in futurer versions?
or
Is this more a Linq related problem/question?

### Further technical details
EF Core version: 2.2.0
Database Provider: any

**Comment by SteffenMangold** [comment]
> What are you trying to cache exactly? - The object. Where it is coming from?

The goal is to cache the query result. Therefore a Hash for the query is needed to identify unique query and cache the result. A popular implementation of this is [EFSesondLevelCaching](https://github.com/VahidN/EFSecondLevelCache.Core). As described above generating a hash of a query is not quite easy.

**Comment by smitpatel** [comment]
Just use the expression tree.

**Comment by SteffenMangold** [comment]
> Just use the expression tree.

Sorry don't get it. Can you give some more details what you mean?

**Comment by smitpatel** [comment]
IQueryable.Expression, just use the whole structure. EF Core uses it for caching. I believe you can also use the same. Sample implementation is here https://github.com/aspnet/EntityFrameworkCore/blob/master/src/EFCore/Query/Internal/ExpressionEqualityComparer.cs
You may need to tweak it a little as it may have something EF Core specific too.

**Comment by SteffenMangold** [comment]
> IQueryable.Expression, just use the whole structure. EF Core uses it for caching. I believe you can also use the same. Sample implementation is here https://github.com/aspnet/EntityFrameworkCore/blob/master/src/EFCore/Query/Internal/ExpressionEqualityComparer.cs
> You may need to tweak it a little as it may have something EF Core specific too.

OK I tried it but it has the same problem.
Here an example:

```c#
var expressionComparer = new ExpressionEqualityComparer();

var myObject = new MyObject
     {
         ID = 1
     };

var bookQuery = model.Books
      .Where(d => d.ID == myObject.ID);

var hashForOneQuery = = expressionComparer.GetHashCode(expression);

myObject.ID = 10;

var hashForTenQuery = = expressionComparer.GetHashCode(expression);
```

Both hash values are the same but the accutual excecuted query differs in result.
Have you read [this artical](https://www.c-sharpcorner.com/article/how-to-enable-c-sharp-7-1-version-to-my-projects/)? I think it descipted what happens.

**Comment by smitpatel** [comment]
Yes, I read that article. But that happens due to lack of GetHashCode/Equals on the closure type or any underlying type except scalar types. Since either it's same reference or it's different reference regardless of underlying value.

So I suggested you to do some tweaking, like you can find out member access on closure (it will be member access always), to scalar and evaluate that and use it in computation of HashCode/Equals.
Following code shows how to get value of closure member efficiently,
https://github.com/aspnet/EntityFrameworkCore/blob/0dfb5f2902ca121db76502085f09ea5230a881ec/src/EFCore/Query/ExpressionVisitors/Internal/ParameterExtractingExpressionVisitor.cs#L296-L320

Though closure variables are essential for EF Core to produce parameters & query caching (on very first level), hence evaluating those expressions in advance is no use. You could also intercept EF Core's query compiler and inject caching mechanism after parameterization EF Core do and compare that way, Effectively, get already compiled query out of EF Core cache, (so expression tree match is already checked) and then compare parameter values and use it to figure out if you want to execute the query or give results from cache.

**Comment by smitpatel** [comment]
I don't think, EF Core should be providing any text representation for caching. We provider Expression Printer for pretty print debug results but not production use. It is the can of worms and leads us to path of bugs. Expression tree itself should be robust enough.

We have sufficient hooks in place for others to intercept expression and inject caching as they see fitting.
Pre-funcletization, user has to deal with closure variables themselves. Post-Functlization, they need to check parameter values too.

Closing this as there is nothing actionable on EF Core side here.

**Comment by SteffenMangold** [comment]
> intercept expression

Thanks, I can understand this. Makes sense for me.
Do you have a link for a good example of intercepting expressions in EF core?
Need to dive a little deeper to solve this. :)

**Comment by SteffenMangold** [comment]
@smitpatel you wrote

> You could also intercept EF Core's query compiler and inject caching mechanism after parameterization EF Core do and compare that way, Effectively, get already compiled query out of EF Core cache, (so expression tree match is already checked) and then compare parameter values and use it to figure out if you want to execute the query or give results from cache.

It's really hard to follow this without have a good documentation. I looks into the EF Core source and try to understand what you mean.

So I created a custom `QueryCompiler` and override `Execute`.

```
 public override TResult Execute<TResult>(Expression query)
        {
            Check.NotNull(query, nameof(query));

            var queryContext = _queryContextFactory.Create();

            //var cacheableOperators = _queryModelGenerator.ParseQuery(query).ResultOperators
            //    .OfType<CacheableResultOperator>()
            //    .ToList();

            query = _queryModelGenerator.ExtractParameters(_logger, query, queryContext);           

            var cacheKey = _compiledQueryCacheKeyGenerator.GenerateCacheKey(query, false);
            var compiledQuery = _compiledQueryCache.GetOrAddQuery(cacheKey, () => CreateCompiledQuery<TResult>(query)); 

            return compiledQuery(queryContext);
        }
```

But I couldnt figure out how to  

> and then compare parameter values and use it 

Because the compiledQuery is simply a function without any possibilty to recieve parameter.

So I ended up with the idea of creating a custom `ExpressionVisitor` and resolve all MemberExppressions. But I realy not understand if this is a proper way. The EF Core logic is realy hard to understand.

**Comment by smitpatel** [comment]
In above custom `QueryCompiler` basically, what you want to do is, 
After this line `query = _queryModelGenerator.ExtractParameters(_logger, query, queryContext);`, you have all closure variables converted to parameters which are available at `queryContext.ParameterValues` so now you can compare the expression tree (since it won't cause closure issues) and also compare `queryContext.ParameterValues` to see if it is same expression tree with same closure values. If it is then don't try to compile the query.

Also the custom result operator you are adding in #14242 , you should not be adding. You need custom result operator only when you want to do something at compile time. From my understanding, you just want to serve from cache if the query had run in the past.
